types/rest/rest.go

```diff
@@ -0,0 +1,241 @@
+	from, password, memo, chainID string, gas, gasAdjustment string,
+	accNumber, seq uint64, fees types.Coins, gasPrices types.DecCoins, genOnly, simulate bool,
+) BaseReq {
+
+	return BaseReq{
+		From:          strings.TrimSpace(from),
+		Password:      password,
+		Memo:          strings.TrimSpace(memo),
+		ChainID:       strings.TrimSpace(chainID),
+		Fees:          fees,
+		GasPrices:     gasPrices,
+		Gas:           strings.TrimSpace(gas),
+		GasAdjustment: strings.TrimSpace(gasAdjustment),
+		AccountNumber: accNumber,
+		Sequence:      seq,
+		GenerateOnly:  genOnly,
+		Simulate:      simulate,
+	}
+}
+
+// Sanitize performs basic sanitization on a BaseReq object.
+func (br BaseReq) Sanitize() BaseReq {
+	return NewBaseReq(
+		br.From, br.Password, br.Memo, br.ChainID, br.Gas, br.GasAdjustment,
+		br.AccountNumber, br.Sequence, br.Fees, br.GasPrices, br.GenerateOnly, br.Simulate,
+	)
+}
+
+// ValidateBasic performs basic validation of a BaseReq. If custom validation
+// logic is needed, the implementing request handler should perform those
+// checks manually.
+func (br BaseReq) ValidateBasic(w http.ResponseWriter) bool {
+	if !br.GenerateOnly && !br.Simulate {
+		switch {
+		case len(br.Password) == 0:
+			WriteErrorResponse(w, http.StatusUnauthorized, "password required but not specified")
+			return false
+
+		case len(br.ChainID) == 0:
+			WriteErrorResponse(w, http.StatusUnauthorized, "chain-id required but not specified")
+			return false
+
+		case !br.Fees.IsZero() && !br.GasPrices.IsZero():
+			// both fees and gas prices were provided
+			WriteErrorResponse(w, http.StatusBadRequest, "cannot provide both fees and gas prices")
+			return false
+
+		case !br.Fees.IsValid() && !br.GasPrices.IsValid():
+			// neither fees or gas prices were provided
+			WriteErrorResponse(w, http.StatusPaymentRequired, "invalid fees or gas prices provided")
+			return false
+		}
+	}
+
+	if len(br.From) == 0 {
+		WriteErrorResponse(w, http.StatusUnauthorized, "name or address required but not specified")
+		return false
+	}
+
+	return true
+}
+
+/*
+ReadRESTReq is a simple convenience wrapper that reads the body and
+unmarshals to the req interface. Returns false if errors occurred.
+
+  Usage:
+    type SomeReq struct {
+      BaseReq            `json:"base_req"`
+      CustomField string `json:"custom_field"`
+		}
+
+    req := new(SomeReq)
+    if ok := ReadRESTReq(w, r, cdc, req); !ok {
+        return
+    }
+*/
+func ReadRESTReq(w http.ResponseWriter, r *http.Request, cdc *codec.Codec, req interface{}) bool {
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		WriteErrorResponse(w, http.StatusBadRequest, err.Error())
+		return false
+	}
+
+	err = cdc.UnmarshalJSON(body, req)
+	if err != nil {
+		WriteErrorResponse(w, http.StatusBadRequest, fmt.Sprintf("failed to decode JSON payload: %s", err))
+		return false
+	}
+
+	return true
+}
+
+// AddrSeed combines an Address with the mnemonic of the private key to that address
+type AddrSeed struct {
```

Comment:
Where is this used?

Reply:
It's used in LCD's `test_helpers.go`. I relocated it in there.